File: 2016/20.py

import math, copy, re, hashlib


def part_1(data):

	data = [[int(i.split('-')[0]), int(i.split('-')[1])] for i in data]
	data.sort(key=lambda x: x[0])
	mini = 0
	for i in data:
		if mini < i[0]:
			break
		if mini >= i[0] and mini <= i[1]:
			mini = i[1] + 1

	print(mini)
	print('END OF PART1')
	return

def merge(l):
	merged = -1
	ret = []

	i = 0
	while i < len(l):
		[a, b] = l[i]
		j = i + 1
		while j < len(l) and l[j][0] < b: 
			b = max(b, l[j][1])
			j += 1
		ret.append([a, b])
		i = j

	return ret

def part_2(data):

	data = [[int(i.split('-')[0]), int(i.split('-')[1])] for i in data]
	data.sort(key=lambda x: x[0])
	cnt = 0
	new_data = merge(data)
	new_data = merge(new_data)
	new_data = merge(new_data)
	new_data = merge(new_data)
	new_data = merge(new_data)
	print(len(data), len(new_data))
	MAX = 4294967295
	for i in range(len(new_data) - 1):
		cnt += new_data[i + 1][0] - new_data[i][1] - 1
	cnt += MAX - new_data[-1][1] if MAX != new_data[-1][1] else 0
	print(cnt)
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('20_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2023/20.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

function part1(data, part = 1) {

	let res = 0;
	data = lines(data);
	const cfg = {};
	iter(data, l => {
		let [modul, out] = l.split('->').map(a => a.trim());
		out = out.split(', ');
		const mode = modul[0];
		modul = mode === 'b' ? modul : modul.slice(1);
		cfg[modul] = { mode, out, mem: null, v: {}, cycle: -1 };
	});
	iter(keys(cfg), k => {
		if (cfg[k].mode !== '&') {
			cfg[k].state = false;
			return;
		}
		const input = {};
		iter(keys(cfg), kk => {
			if (cfg[kk].out.indexOf(k) !== -1) {
				input[kk] = false;
			}
		});
		cfg[k].mem = input;
		cfg[k].state = true;
	})
	let l = 0, h = 0;
	let cnt = 0;
	let done = false
	const RES = { sg: null, lm: null, dh: null, db: null };
	let resCnt = 0;
	while (!done) {
		const q = new Queue();
		q.push('broadcaster');
		++l;
		while (!done && !q.empty()) {
			const s = q.pop();
			const curr = cfg[s].state;
			iter(cfg[s].out, k => {
				if (curr) ++h;
				else ++l;
				if (s in RES && RES[s] === null && curr) {
					RES[s] = cnt + 1;
					++resCnt;
				}
				if (part === 2 && resCnt === 4) {
					res = prod(keys(RES).map(k => RES[k]));
					done = true;
					return done;
				}
				if (!(k in cfg)) return;
				if (cfg[k].mode === '%') {
					if (curr) return;
					cfg[k].state = !cfg[k].state;
				} else {
					cfg[k].mem[s] = curr;
					cfg[k].state = !keys(cfg[k].mem).every(ii => cfg[k].mem[ii]);
				}
				q.push(k);
			})
		}
		if (part === 1 && cnt === 999) {
			res = h * l;
			done = true;
		}
		++cnt;
	}

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

function part2(data) {

	let res = 0;

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("20_input").toString("utf-8");

	part1(data);
	part1(data, 2);
	process.exit(0);
}

main();
